src/main/java/com/tvd12/gamebox/entity/MMOGridRoom.java

```diff
@@ -0,0 +1,235 @@
+        final int cellOfInterestEndY = Math.min(cells.length - 1, cell.cellY + cellRangeOfInterest);
+        final int cellOfInterestEndZ = Math.min(cells.length - 1, cell.cellZ + cellRangeOfInterest);
+        final int cellOfInterestStartY = Math.max(0, cell.cellY - cellRangeOfInterest);
+        final int cellOfInterestStartZ = Math.max(0, cell.cellZ - cellRangeOfInterest);
+        
+        cellPlayerBuffer.clear();
+        cell.getPlayerList(cellPlayerBuffer);
+        
+        for (int i = 0; i < cellPlayerBuffer.size(); ++i) {
+            MMOPlayer currentPlayer = cellPlayerBuffer.get(i);
+            
+            // Handle players in the current cell
+            for (int j = i; j < cellPlayerBuffer.size(); ++j) {
+                currentPlayer.addNearbyPlayer(cellPlayerBuffer.get(j));
+                cellPlayerBuffer.get(j).addNearbyPlayer(currentPlayer);
+            }
+            
+            // Handle players in neighboring cells
+            for (int ix = cell.cellX + 1; ix <= cellOfInterestEndX; ++ix) {
+                for (int iy = cellOfInterestStartY; iy <= cellOfInterestEndY; ++iy) {
+                    for (int iz = cellOfInterestStartZ; iz <= cellOfInterestEndZ; ++iz) {
+                        addNearbyPlayersInCell(currentPlayer, cells[ix][iy][iz]);
+                    }
+                }
+            }
+            for (int iy = cell.cellY + 1; iy <= cellOfInterestEndY; ++iy) {
+                for (int iz = cellOfInterestStartZ; iz <= cellOfInterestEndZ; ++iz) {
+                    addNearbyPlayersInCell(currentPlayer, cells[cell.cellX][iy][iz]);
+                }
+            }
+            for (int iz = cellOfInterestStartZ; iz <= cellOfInterestEndZ; ++iz) {
+                if (iz != cell.cellZ) {
+                    addNearbyPlayersInCell(currentPlayer, cells[cell.cellX][cell.cellY][iz]);
+                }
+            }
+        }
+    }
+    
+    public void addNearbyPlayersInCell(MMOPlayer currentPlayer, Cell cell) {
+        if (cell == null || cell.getNumberOfPlayers() == 0) {
+            return;
+        }
+        Collection<MMOPlayer> players = cell.getPlayerCollection();
+        for (MMOPlayer otherPlayer : players) {
+            currentPlayer.addNearbyPlayer(otherPlayer);
+            otherPlayer.addNearbyPlayer(currentPlayer);
+        } 
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static class Builder extends MMORoom.Builder {
+
+        private int maxX;
+        private int maxY;
+        private int maxZ;
+        private int cellSize;
+
+        public Builder maxX(int maxX) {
+            this.maxX = maxX;
+            return this;
+        }
+
+        public Builder maxY(int maxY) {
+            this.maxY = maxY;
+            return this;
+        }
+
+        public Builder maxZ(int height) {
+            this.maxZ = height;
+            return this;
+        }
```

Comment:
Ý nghĩa của các trường này là max number of cells hay là độ rộng của room theo x, y, z em nhỉ? nếu là max number of cells thì đặt tên này có vẻ gây nhầm lẫn, nên là numberOfCellsX, numberOfCellsY, numberOfCellsZ có vẻ rõ ràng hơn em nhỉ?

Reply:
Các trường này là độ rộng của room theo x, y, z.

Em thấy để vậy tiện cho dev hơn vì:
- Cách nào dev cũng phải tự tính ra cell size
- Nếu dùng numberOfCellsX thì đơn vị của nó bị khác đơn vị của cell size, và numberOfCellsX cũng khó xác định hơn cho dev do các game engine cho ra tọa độ x, y, z chứ không cho ra số cells